packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+            _subject,
+            _amount,
+            _lockPeriodInSec,
+            _beneficiary,
+            false
+        );
+
+        // Transfer the tokens to this contract
+        subjectToken.safeTransferFrom(msg.sender, address(this), _amount);
+    }
+
+    /**
+     * @notice Sets the lock period for staking. only owner can call this function.
+     * @param _lockPeriodInSec Lock period to set.
+     * @param _allowed Boolean to allow or disallow the lock period.
+     */
+    function setLockPeriod(
+        uint256 _lockPeriodInSec,
+        bool _allowed
+    ) external onlyRole(CHANGE_LOCK_DURATION) {
+        if (lockPeriodsInSec[_lockPeriodInSec] == _allowed) {
+            revert Staking_LockPeriodAlreadySet();
+        }
+        lockPeriodsInSec[_lockPeriodInSec] = _allowed;
+        emit LockPeriodUpdated(_lockPeriodInSec, _allowed);
+    }
+
+    /**
+     * External function to deposit and lock tokens.
+     * @param _subject Subject address for which tokens are getting deposited.
+     * @param _amount amount of tokens getting deposited.
+     * @param _lockPeriodInSec lock period for the tokens.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function depositAndLock(
+        address _subject,
+        uint256 _amount,
+        uint256 _lockPeriodInSec
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 unlockTimeInSec_)
+    {
+        unlockTimeInSec_ = _depositAndLock(
+            _subject,
+            _amount,
+            _lockPeriodInSec,
+            msg.sender
+        );
+    }
+
+    /**
+     * External function to deposit and lock tokens for a beneficiary.
+     * @param _subject Subject address for which tokens are getting deposited.
+     * @param _amount amount of tokens getting deposited.
+     * @param _lockPeriodInSec lock period for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function depositAndLockFor(
+        address _subject,
+        uint256 _amount,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 unlockTimeInSec_)
+    {
+        unlockTimeInSec_ = _depositAndLock(
+            _subject,
+            _amount,
+            _lockPeriodInSec,
+            _beneficiary
+        );
+    }
+
+    /**
+     * External function to deposit and lock multiple tokens.
+     * @param _subjects Subject addresses for which tokens are getting deposited.
+     * @param _amounts Amounts of tokens getting deposited.
+     * @param _lockPeriodInSec Lock periods for the tokens.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function depositAndLockMultiple(
+        address[] memory _subjects,
+        uint256[] memory _amounts,
+        uint256 _lockPeriodInSec
```

Comment:
should this be an array as well. user can configure different lock periods for different subjects.

Reply:
No, all locks are for a given period to keep implementation simple.